build/generate-licenses.go

```diff
@@ -115,8 +126,147 @@ func main() {
 			log.Fatalf("Failed to write new file. %s", err)
 		} else {
 			fmt.Printf("Written %s\n", out.Name())
+
+			// some license files have same content, so we need to detect these files and create a convert map into a file
+			// In InitClassifier, we will use this convert map to avoid adding same license content with different license name
+			md5, err := getSameFileMD5(preFile, out)
+			if err != nil {
+				log.Fatalf("Failed to get same file md5. %s", err)
+				continue
+			}
+			if md5 != "" {
+				_, ok := sameFiles[md5]
+				if !ok {
+					sameFiles[md5] = make([]string, 0)
+				}
+				if !slices.Contains(sameFiles[md5], preLicenseName) {
+					sameFiles[md5] = append(sameFiles[md5], preLicenseName)
+				}
+				sameFiles[md5] = append(sameFiles[md5], licenseName)
+			}
+			preFile = out
+			preLicenseName = licenseName
+		}
+	}
+
+	// generate convert license name map
+	sameLicenses := make(map[string]string)
+	for _, fileNames := range sameFiles {
+		key := getLicenseKey(fileNames)
+		for _, fileName := range fileNames {
+			sameLicenses[fileName] = key
 		}
 	}
+	// save convert license name map to file
+	bytes, err := json.Marshal(sameLicenses)
+	if err != nil {
+		log.Fatalf("Failed to create json bytes. %s", err)
+		return
+	}
+	out, err := os.Create(sameLicensesDestination)
+	if err != nil {
+		log.Fatalf("Failed to create new file. %s", err)
+	}
+	defer out.Close()
+	_, err = out.Write(bytes)
+	if err != nil {
+		log.Fatalf("Failed to write same licenses json file. %s", err)
+	}
 
 	fmt.Println("Done")
 }
+
+// getSameFileMD5 returns md5 of the input file, if the content of input files are same
+func getSameFileMD5(f1, f2 *os.File) (string, error) {
```

Comment:
I do not think it needs such function.

You could just compare the files byte by byte.

That's an over-engineering to use MD5 here.

Reply:
I think the hash is ok because then you must read a file only once. But the current method is bad designed. Every tar entry is written to disc, read again and read again in the next iteration. The `io.Copy(out, tr)` in line 125 could create a hash on the fly without re-reading a file.

Reply:
If you could compare a file byte-by-byte, then why do extra work / use more CPU to calculate the hash?

Reply:
The code compares a file with all other files. If you would compare them byte-by-byte you would have to read every file n times. With the hash you only need to read a file once.

Reply:
OK, I think I understand the problem now and I see how the "md5" works here. But the code really reads strange, still over-engineering IMO .....